src/adapter/src/session.rs

```diff
@@ -829,14 +883,84 @@ pub enum TransactionOps<T> {
     /// is Some, it must only do other peeks. However, if the value is None
     /// (i.e. the values are constants), the transaction can still perform
     /// writes.
-    Peeks(Option<(T, Option<Timeline>)>),
+    Peeks(ReadContext<T>),
     /// This transaction has done a `SUBSCRIBE` and must do nothing else.
     Subscribe,
     /// This transaction has had a write (`INSERT`, `UPDATE`, `DELETE`) and must
     /// only do other writes, or reads whose timestamp is None (i.e. constants).
     Writes(Vec<WriteOp>),
 }
 
+impl<T> TransactionOps<T> {
+    fn read_context(self) -> Option<ReadContext<T>> {
+        match self {
+            TransactionOps::Peeks(read_context) => Some(read_context),
+            TransactionOps::None | TransactionOps::Subscribe | TransactionOps::Writes(_) => None,
+        }
+    }
+}
+
+impl<T> Default for TransactionOps<T> {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+/// The timeline and timestamp context of a read.
+#[derive(Debug, Clone, PartialEq)]
+pub enum ReadContext<T> {
+    /// Read is executed in a specific timeline with a specific timestamp.
+    TimelineTimestamp(Timeline, T),
+    /// Read is execute without a timeline or timestamp.
+    NoTimestamp,
+}
+
+impl<T> ReadContext<T> {
+    /// Creates a `ReadContext` from a timestamp and `TimelineContext`.
+    pub fn from_timeline_context(
+        ts: T,
+        timeline: Option<Timeline>,
+        timeline_context: TimelineContext,
+    ) -> Result<ReadContext<T>, AdapterError> {
+        match timeline_context {
+            TimelineContext::TimelineDependent(timeline) => {
+                Ok(Self::TimelineTimestamp(timeline, ts))
```

Comment:
Is it safe to add:

```
assert_eq!(timeline, transaction_timeline);
```

Reply:
It would have to be:
```
if let Some(transaction_timeline) = transaction_timeline {
  assert_eq!(timeline, transaction_timeline);
}
```
so it doesn't panic on the first query of a transaction. I'll add this.

Reply:
Added this.